### PR TITLE
Variable support

### PR DESCRIPTION
**Pull request by thijs-obj**
# End User Changes:
Variable support. Variables are created as Series which can be used similar to constants, except that their value can later be changed. Variables are identified by a combination of name and dtype.

Specifically the following new functions are available:
  * `DataFrame.create_variable()`: adds the variable and its initial value to `self.variables` and returns a `Series` that can be used as a placeholder whose value can later be changed. Wherever the returned `Series` object is used, the value of the variable is set when the state of the DataFrame is translated to a sql query
  * `DataFrame.set_variable()`. Set a value in `self.variables`
  * `DataFrame.get_all_variable_usage()`: Give an overview of all variables that are used. This can help with diagnosing variable problems.

# Code Changes:
* Supporting change: Introduce two `BachSqlModel` subclasses: `CurrentNodeSqlModel` and `MergeSqlModel`.
  * The `__init__` of these classes generate the raw sql.
  * Models have no properties that are `Expressions` or `Lists[Expression]`. All Expressions are translated to sql in the Model `__init__()`
    *  `Expression.to_sql()` now assumes it will become raw Model.sql and will have format() called on in twice
  *  Got rid of `BachSqlModelBuilder`
* Add ExpressionToken `VariableToken` that can be used as a placeholder for values
* `DataFrame` has a mapping `variables` with values for variables
* `df.variables` are used and set as properties when creating instances of `CurrentNodeSqlModel` and `MergeSqlModel`.
* `df.variables` are used to replace existing properties in the whole tree of `self.base_node`

### CODE REVIEW

File: bach/tests/functional/bach/test_df_variable_expression.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        df,
+        expected_columns=['_index_skating_order', 'city', 'founding'],
+        expected_data=[
+            [1, 'Ljouwert', 1285],
+            [2, 'Snits', 1456],
+            [3, 'Drylts', 1268],
+        ]
+    )
+
+    df, add_value = df.create_placeholder(name='add_value', value=1000)
```
Comment (thijs-obj) [inline bach/tests/functional/bach/test_df_variable_expression.py:20]: This test shows the main workflow:
- Use `create_placeholder` to add a placeholder, which returns a copy of the dataframe with the placeholder, and a series that acts as the placeholder
- Later on `df.set_placeholder` can be used to change the value of the placeholder

Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_variable_expression.py:20]: I called these things placeholder for now, but maybe we want to call it variable

File: bach/bach/dataframe.py [hunk 2]
```diff
@@ -1451,38 +1470,44 @@ def get_current_node(self, name: str,
             having_clause = having_clause if having_clause else Expression.construct('')
 
             columns += [s.get_column_expression() for s in self._data.values()]
+            # TODO: this breaks a lot of injection escaping logic
```
Comment (thijs-obj) [inline bach/bach/dataframe.py:1473]: This is one of the many reasons why this is work-in-progress

File: bach/tests/functional/bach/test_df_variable_expression.py (hunk 1 above)
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_variable_expression.py:20]: Maybe good to review this test already. Is this sort of what we had in mind when we said 'Variable support' in our shared document?
 @jansentom

File: bach/sql_models/model.py [hunk 3]
```diff
@@ -300,11 +297,10 @@ def set_values(self: TB, **values) -> TB:
                     raise ValueError(f'reference ({key}) is of incorrect type: {type(value)}')
                 self._references[key] = value
             elif key in self.spec_properties:
-                # todo: enable below check, after we stop using lists in properties in Bach
```
Comment (thijs-obj) [inline bach/sql_models/model.py:303]: BachSqlModelBuilder was retired, and we now translate `List[Expression]` to sql when we create the SqlModel instead of when we generate the sql. So we don't need to be able to have properties that are mutable (as the lists were)

File: bach/bach/dataframe.py [hunk 4]
```diff
@@ -2003,6 +2029,103 @@ def var(self, axis=1, skipna=True, level=None, ddof: int = 1, numeric_only=False
...
                                     skipna=skipna, ddof=ddof, **kwargs)
 
+    def create_variable(self, name: str, value: Any) -> Tuple['DataFrame', 'Series']:
+        """
+        Create a Series object that can be used as a variable, within the returned DataFrame. The
+        DataFrame will have the variable with the given values set in :meth:`DataFrame.variables`.
+
+        The variable value can later be changed using :meth:`DataFrame.set_variable`
+
+        ## Multiple variables with the same name
```
Comment (thijs-obj) [inline bach/bach/dataframe.py:2039]: This might be interesting part to read

File: bach/bach/dataframe.py [hunk 5]
```diff
@@ -2003,6 +2029,103 @@ def var(self, axis=1, skipna=True, level=None, ddof: int = 1, numeric_only=False
...
+        return df, series
+
+    def set_variable(self, name: str, value: Any) -> 'DataFrame':
+        """
+        Return a copy of this DataFrame with the variable value updated.
+        """
+        df, _ = self.create_variable(name, value)
+        return df
+
+    def get_all_variable_usage(self) -> List[DefinedVariable]:
```
Comment (thijs-obj) [inline bach/bach/dataframe.py:2062]: This function is not directly useful for any workflow other than debugging. But seemed good to have to me, as I can imagine that these variable can lead to subtle bugs when variables get re-used

File: bach/sql_models/model.py [hunk 6]
```diff
@@ -669,3 +665,18 @@ def sql(self):
     @property
     def generic_name(self):
         return self._generic_name
+
+
+def escape_format_string(value: str, times=1) -> str:
```
Comment (jansentom) [inline bach/sql_models/model.py:670]: I find this really confusing, shouldn't we just split this into two methods?

File: bach/tests/functional/bach/test_df_variables.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            [1, 'Ljouwert sted', 4285],
+            [3, 'Drylts sted', 4268],
+        ]
+    )
+
+
+def test_set_variable_double_types():
+    # It's possible to have two variables with the same name but different type.
+    # This seems counter-intuitive, but removes a lot of the potential edge cases and problems around merging
+    # DataFrames and Series with the same variable names.
```
Comment (jansentom) [inline bach/tests/functional/bach/test_df_variables.py:82]: Oi, this is going to cause a lot of confusing debug session. Are we sure we want to keep this? Can you ever update this variable by name?

File: bach/tests/functional/bach/test_injection_column_name_escaping.py [hunk 8]
```diff
@@ -27,22 +23,30 @@ def test_column_names():
...
+
+
+def _get_dataframe_with_weird_column_names():
     bt = get_bt_with_test_data()[['city']]
     bt['With_Capitals'] = 1
     bt['with_capitals'] = 1
     bt['With A Space Too'] = 1
     bt['""with"_quotes""'] = 1
+    bt['with%percentage'] = 1
+    bt['with{format}{{strings}}{{}'] = 1
```
Comment (jansentom) [inline bach/tests/functional/bach/test_injection_column_name_escaping.py:48]: :)

File: bach/tests/functional/bach/test_savepoints.py [hunk 9]
```diff
@@ -91,9 +91,9 @@ def test_write_to_db_create_objects():
     expected_columns = ['_index_skating_order', 'skating_order', 'city', 'founding', 'x']
     expected_data = [[1, 1, 'Ljouwert', 1285, 12345]]
 
-    assert sps.to_sql()['sp_final_point'] == \
+    assert sps.to_sql()['sp_final_point'].replace('\n', '') == \
```
Comment (jansentom) [inline bach/tests/functional/bach/test_savepoints.py:94]: re.match() for robustness?

File: bach/bach/merge.py [hunk 10]
```diff
@@ -331,3 +334,55 @@ def _get_expression(df_series: DataFrameOrSeries, label: str) -> Expression:
     if isinstance(df_series, Series):
         return df_series.expression
     raise TypeError(f'df_series should be DataFrameOrSeries. type: {type(df_series)}')
+
+
+class MergeSqlModel(BachSqlModel):
```
Comment (jansentom) [inline bach/bach/merge.py:339]: nice to abstract this like this.

File: bach/bach/merge.py [hunk 11]
```diff
@@ -331,3 +334,55 @@ def _get_expression(df_series: DataFrameOrSeries, label: str) -> Expression:
...
+        :param right_node, sql-model of the materialized right side of the join
+        :param variables: Dictionary of all variable values
+        """
+        columns_str = columns_expr.to_sql()
+        join_type_str = join_type_expr.to_sql()
+        on_str = on_clause.to_sql()
+
+        sql = f'''
+            select {columns_str}
+            from {{{{left_node}}}} as l {join_type_str}
```
Comment (jansentom) [inline bach/bach/merge.py:365]: this looks quite nasty now :( Are there alternatives?

**Review by jansentom — APPROVED** [review]
Very clean PR and with very useful intro. it was quite a pleasure reviewing. :100: 

Some small, mostly aesthetic comment, but good to go from my end!

File: bach/tests/functional/bach/test_savepoints.py (hunk 9 above)
Comment (thijs-obj) [reply bach/tests/functional/bach/test_savepoints.py:94]: Tried something like
```python
    assert re.match(
        'select "_index_skating_order", "skating_order", "city", "founding", "x".*from.*"sp_third_point".*limit all.*',
        sps.to_sql()['sp_final_point'], flags=re.DOTALL
    )
```

But I find the readability not much better, and debugging it is actually more annoying because the error is less clear

File: bach/tests/functional/bach/test_df_variables.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    with pytest.raises(Exception, match='Variable int_variable, with dtype int64 is used, but not set.'):
+        df_without_variable.to_pandas()
+
+    df.materialize(inplace=True)
+    # Now that the variable is not used in the series anymore, only in base_node we should be able to unset
+    # it.
+    df_without_variable = df.copy_override(variables={})
+    assert df_without_variable.variables == {}
+    # now this should not raise an error
+    df_without_variable.to_pandas()
```
Comment (hendrik-obj) [inline bach/tests/functional/bach/test_df_variables.py:125]: There is no variable in `df_without_variable`. However `df_without_variable.get_all_variable_usage()` says `int_variable` is used, with an old_value, but without a ref_path (which I would expect).

Comment (hendrik-obj) [reply bach/tests/functional/bach/test_df_variables.py:125]: Also if you do `df_without_variable, int_variable = df_without_variable.create_variable('int_variable', 4000)` the variable before materialization is modified to 4000.
However if you would have done `df_without_variable = df.copy_override()` the variable can't be set with `set_variable` (`df_without_variable.set_variable(name='int_variable', value=4000)` doesn't change the varable in `df_without_variable.variables`)

**Review by hendrik-obj — COMMENTED** [review]
nice, but complex stuff. Some stuff confuses me a bit, mostly revolving around materialization.

File: bach/sql_models/model.py (hunk 6 above)
Comment (thijs-obj) [reply bach/sql_models/model.py:670]: Discussed out of band: Will create two separate functions: `escape_raw_sql` and `escape_placeholder` or something similarly named

File: bach/tests/functional/bach/test_df_variables.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+        DtypeNamePair(dtype='string', name='shared'): ' first',
+        DtypeNamePair(dtype='string', name='variable1'): ' city'
+    }
+    assert mt.variables == {
+        DtypeNamePair(dtype='string', name='shared'): ' second',
+        DtypeNamePair(dtype='string', name='variable2'): ' food'
+    }
+
+    result = bt.merge(mt)
+    # The 'shared' variable is in both DataFrames. The value from the left df will override the right
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_variables.py:149]: Good point. Added documentation: https://github.com/objectiv/objectiv-analytics/pull/356/commits/56d1489032799b21e9dc29eab191d569a7a162ee

File: bach/tests/functional/bach/test_df_variables.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+        df,
+        expected_columns=['_index_skating_order', 'city', 'founding'],
+        expected_data=[
+            [1, 'Ljouwert city', 3285],
+            [2, 'Snits city', 3456],
+            [3, 'Drylts city', 3268],
+        ]
+    )
+
+    df = df.materialize()
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_variables.py:61]: Discussed out of band.

File: bach/bach/dataframe.py [hunk 15]
```diff
@@ -2003,6 +2029,103 @@ def var(self, axis=1, skipna=True, level=None, ddof: int = 1, numeric_only=False
         return self._aggregate_func('var', axis, level, numeric_only,
                                     skipna=skipna, ddof=ddof, **kwargs)
 
+    def create_variable(self, name: str, value: Any) -> Tuple['DataFrame', 'Series']:
+        """
+        Create a Series object that can be used as a variable, within the returned DataFrame. The
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/bach/dataframe.py:2042]: Discussed out of band. Conclusion was that we should have a 'howto' document or something similar that explains the concept of variables to users

File: bach/bach/dataframe.py [hunk 16]
```diff
@@ -2003,6 +2029,103 @@ def var(self, axis=1, skipna=True, level=None, ddof: int = 1, numeric_only=False
...
+        edge cases around merging DataFrames and Series with the same variables, and building on top of
+        SqlModels that already have variables.
+
+        :return: Tuple with DataFrame and Series object.
+        """
+        from bach.series.series import variable_series
+        series = variable_series(base=self, value=value, name=name)
+        variables = self.variables
+        variables[DtypeNamePair(dtype=series.dtype, name=name)] = value
+        df = self.copy_override(variables=variables)
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/bach/dataframe.py:2066]: discussed this. Will keep as is

File: bach/tests/functional/bach/test_df_variables.py (hunk 12 above)
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_variables.py:125]: discussed this. Will keep as is

File: bach/bach/merge.py (hunk 11 above)
Comment (thijs-obj) [reply bach/bach/merge.py:365]: discussed. Will keep as is

File: bach/tests/functional/bach/test_df_variables.py (hunk 7 above)
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_variables.py:82]: Discussed this out of band. Will keep this as

File: bach/sql_models/model.py (hunk 6 above)
Comment (thijs-obj) [reply bach/sql_models/model.py:686]: See https://github.com/objectiv/objectiv-analytics/pull/356/commits/56d1489032799b21e9dc29eab191d569a7a162ee

File: bach/bach/dataframe.py [hunk 17]
```diff
@@ -2029,7 +2032,12 @@ def var(self, axis=1, skipna=True, level=None, ddof: int = 1, numeric_only=False
         return self._aggregate_func('var', axis, level, numeric_only,
                                     skipna=skipna, ddof=ddof, **kwargs)
 
-    def create_variable(self, name: str, value: Any) -> Tuple['DataFrame', 'Series']:
+    def create_variable(
+            self,
+            name: str,
+            value: Any,
+            *, dtype: Optional[str] = None
```
Comment (thijs-obj) [inline bach/bach/dataframe.py:2039]: Also added the `dtype` as optional parameter here. I think we discussed this at some point @jansentom FYI

**Comment by thijs-obj** [comment]
I'll merge this. If there are any more concerns, please let met know
